### DISCUSSION TITLE
Is it possible to execute bash/shell scripts?

### DISCUSSION THREAD

**Discussion by royfrancis** [quarto-dev/quarto-cli#5275] [Q&A] [answered]
Is it possible to execute bash/shell scripts? Or at least show code highlighting?

````
---
title: Code highlighting
---

# Python

```{python}
import os
os.getcwd()
```

# Shell

```{bash}
#!/bin/bash
echo "Today is " `date`
```
````

![Screenshot from 2023-04-22 19-10-28](https://user-images.githubusercontent.com/5083500/233797576-42a5e50b-ee18-43f7-9c32-4475cdd314ca.png)

**Comment by royfrancis** [comment]
Seems like I have found a solution. Setting `engine: knitr` makes it work. Surely, there must be a solution that doesn't involve R.

````
---
title: Code highlighting
engine: knitr
---

## Python

```{python}
import os
os.getcwd()
```

## Shell

```{bash}
#!/bin/bash
echo "Today is " `date`
```
````

![Screenshot from 2023-04-22 19-29-27](https://user-images.githubusercontent.com/5083500/233798407-4705967f-691e-4d62-8e43-e37a848016e4.png)

**Comment by mcanouil** [comment]
You can use Jupyter (as long as you have the right kernel to execute Bash code) instead of `knitr`, but you need to use an engine.
See #4634 for related discussion.

**Comment by royfrancis** [comment]
@mcanouil  Need to revive this question. Why don't bash chunks work with jupyter engine?

**knitr.qmd**

````
---
title: knitr engine
engine: knitr
---

```{python}
#| filename: python
#| eval: true

from datetime import date
print("Today's date:", date.today())
```

```{bash}
#| filename: bash
#| eval: true

#!/bin/bash
echo "Today is " `date`
```
````

![Screenshot 2023-10-10 at 10 57 25](https://github.com/quarto-dev/quarto-cli/assets/5083500/040045d4-ca0a-494b-85fb-2c97f18ba1cd)

**jupyter.qmd**

````
---
title: jupyter engine
engine: jupyter
---

```{python}
#| filename: python
#| eval: true

from datetime import date
print("Today's date:", date.today())
```

```{bash}
#| filename: bash
#| eval: true

#!/bin/bash
echo "Today is " `date`
```
````

![Screenshot 2023-10-10 at 10 57 34](https://github.com/quarto-dev/quarto-cli/assets/5083500/c3c04288-4906-477f-81fc-9bb92633497f)

```
$quarto --version
1.3.433
```

Related to #7172

**Reply by mcanouil** [reply]
> Why don't bash chunks work with jupyter engine?

That's a question for Jupyter developers.

There are kernels for bash in Jupyter.
This was answered in the various thread linked.

**Reply by cscheid** [reply, marked as answer]
Just to correct a little: it's not a question for the Jupyter developers, because quarto is the tool converting .qmd code cells to .ipynb notebooks.

With that said, it _is_ the case that Jupyter notebooks generally don't support the execution of code cells in multiple languages the way that knitr does. So, with the Jupyter engine, you need to find a different way to do it.

The Python kernel in the Jupyter engine offers cell magics, and that's how we expose shell execution capabilities in quarto: https://quarto.org/docs/computations/execution-options.html#shell-commands
